### PR TITLE
[AOTInductor] Support quantized linear on CPU with fbgemm

### PR DESCRIPTION
**Pull request by muchulee8**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122939

Summary:
Added support for quantized linear on CPU with fbgemm.
Specifically, for torch.ops.quantized.linear_unpacked_dynamic_fp16, we
decompose it into two steps, pack weight, and fbgemm's qlinear with
packed weight.

Test Plan:
Included in commit.
test_aot_inductor::test_quantized_linear

Reviewers:

Subscribers:

Tasks:

Tags:
@diff-train-skip-merge
cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D55512029](https://our.internmc.facebook.com/intern/diff/D55512029)

### CODE REVIEW

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

File: torch/csrc/inductor/aoti_torch/shim_common.cpp [hunk 1]
```diff
@@ -665,6 +667,35 @@ AOTITorchError aoti_torch_mm_out(
...
 }
 
+AOTITorchError aoti_torch_cpu_wrapped_fbgemm_pack_gemm_matrix_fp16(
+    AtenTensorHandle weight,
+    AtenTensorHandle* out) {
+  at::Tensor* weight_tensor = tensor_handle_to_tensor_pointer(weight);
+  auto packed_weight = at::fbgemm_pack_gemm_matrix_fp16(*weight_tensor);
+
+  at::Tensor* out_tensor = new at::Tensor(std::move(packed_weight));
+  *out = tensor_pointer_to_tensor_handle(out_tensor);
```
Comment (chenyang78) [inline torch/csrc/inductor/aoti_torch/shim_common.cpp:677]: Can we use ```new_tensor_handle```?

File: torch/csrc/inductor/aoti_torch/shim_common.cpp [hunk 2]
```diff
@@ -665,6 +667,35 @@ AOTITorchError aoti_torch_mm_out(
...
+    AtenTensorHandle* out) {
+  at::Tensor* input_tensor = tensor_handle_to_tensor_pointer(input);
+  at::Tensor* weight_tensor = tensor_handle_to_tensor_pointer(weight);
+  at::Tensor* bias_tensor = tensor_handle_to_tensor_pointer(bias);
+
+  auto out_result = at::fbgemm_linear_fp16_weight_fp32_activation(
+      *input_tensor, *weight_tensor, *bias_tensor);
+
+  at::Tensor* out_tensor = new at::Tensor(std::move(out_result));
+  *out = tensor_pointer_to_tensor_handle(out_tensor);
```
Comment (chenyang78) [inline torch/csrc/inductor/aoti_torch/shim_common.cpp:695]: Use new_tensor_handle?

**Comment by muchulee8** [comment]
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55512029).

File: torch/csrc/inductor/aoti_torch/shim_common.cpp (hunk 2 above)
Comment (muchulee8) [reply torch/csrc/inductor/aoti_torch/shim_common.cpp]: Done.

File: torch/csrc/inductor/aoti_torch/shim_common.cpp (hunk 1 above)
Comment (muchulee8) [reply torch/csrc/inductor/aoti_torch/shim_common.cpp]: Done.s

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Review by desertfire — APPROVED** [review]

**Pull request closed without merging**